User/pagoe/post build update (#698)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/concepts/security.md

---
description: Understand how to develop secure WebView2 applications
title: Best practices for developing secure WebView2 applications
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 05/21/2020
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, win32 apps, win32, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html, security
---

# Best practices for developing secure WebView2 applications

> When used correctly, hosting web content inside your native application offers certain advantages, such as ___. To avoid vulnerabilities that can arise from hosting web content, make sure to design your WebView2 application to closely monitor interactions between the web content and the host application. 

Reviewer: Can you replace ___ with text please?

Author: There are more than just a few really depends on the use case want to avoid listing things out, doing so would only narrow the amount of use cases

Reviewer: LGTM!

---
description: Learn about how to build a secure WebView2 application
title: WebView2 Security Best Practicies
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 05/21/2020
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, win32 apps, win32, edge, ICoreWebView2, ICoreWebView2Host, browser control, edge html, security
---

# Security Best Practices

The WebView2 control hosts web content. While there are many advantages of hosting web content inside your native application, when used incorrectly, web content can create vulnerabilities. Therefore, we recommend designing your WebView2 appication to closely monitor the interactions between the web content and the host application in line with the following practices.

## Best Practices

1. Treat all web content as insecure

>     - Validate web messages and host object parameters before consuming them, because web messages and parameters can be malformed (unintentionally or maliciously) and cause the app to behave unexpectedly.

Reviewer: ```suggestion
* To prevent using malformed web messages and parameters, validate web messages and host object parameters to prevent your application from behaving unexpectedly. 
```

Reviewer: Can we include a link to an example of how to do this? Or include a code snippet?

Author: Not really a standard way to do this, based on the usage of the developer

>     - Always check the origin of the document running inside WebView2 and assess the trustworthiness of the content. 

Reviewer: ```suggestion
* Assess the trustworthiness of the content by checking the origin of the document running inside WebView2.
```

Reviewer: Can we include a link to an example of how to do this? Or include a code snippet?

Author: Not really a standard way to do this, based on the usage of the developer

> 2.	Design sepcific web message and host object interactions instead of using generic proxies. This will minimize potential vulnerabilities.

Reviewer: ```suggestion
2. Design specific web messages and host object interactions instead of using generic proxies.
```

Reviewer: Can we include a link to an example of how to do this? Or include a code snippet?

> 3. Modify **CoreWebView2Settings** to restrict web content functionality as much as possible

Reviewer: ```suggestion
3. Restrict web content functionality by modifying `CoreWebView2Settings` as follows:
```

>     - Set **AreHostObjectsAllowed** to `false`, if you don’t expect the web content to access host objects

Reviewer: ```suggestion
    - Set `AreHostObjectsAllowed` to `false`, if you don’t expect the web content to access host objects.
```

>     - Set **IsWebMessageEnabled** to `false`, if you don’t expect the web content to post web messages to the app

Reviewer: ```suggestion
    - Set `IsWebMessageEnabled` to `false`, if you don’t expect the web content to post web messages to your native application. 
```

>     - Set **IsScriptEnabled** to `false`, if you don’t expect the web content to run script (i.e. showing static html like EULA)

Reviewer: ```suggestion
    - Set `IsScriptEnabled` to `false`, if you don’t expect the web content to run scripts (for example, you can only show static html content and not run JavaScript in the web content).
```

>     - Set **AreDefaultScriptDialogsEnabled** to `false`, if you don’t expect the web content to show `alert` or `prompt` popups

Reviewer: ```suggestion
    - Set `AreDefaultScriptDialogsEnabled` to `false`, if you don’t expect the web content to show `alert` or `prompt` dialog boxes.
```

> 4.	Use the **NavigationStarting** and **FrameNavigationStarting** events to update settings according to the origin of the new page.

Reviewer: ```suggestion
4.	Use the `NavigationStarting` and `FrameNavigationStarting` events to update settings based on the origin of the new page as follows:
```

>     1.	Block the navigation to pages and frames that you don't want your application to navigate to. 

Reviewer: ```suggestion
    1.	To prevent your application from navigating to certain pages, use these events to check and then block page or frame navigation. 
```

Reviewer: Can we include a link to an example of how to do this? Or include a code snippet?

>     2.	Adjust **CoreWebView2Settings** according to the needs and trustworthiness of the new page in accordance with the principles above.

Reviewer: ```suggestion
    2.	When navigating to a new page, you may need to adjust the property values on `CoreWebView2Settings` as described above.
```

> 5.	Use the **ContentLoading** event, which fires when the WebView2 switches to a new document, to add and remove host objects using **AddHostObjectToScript** and **RemoveHostObjectFromScript** to minimize the number of exposed host objects.

Reviewer: ```suggestion
5. When navigating to a new document, use the `ContentLoading` event to remove exposed host objects using `RemoveHostObjectFromScript`.
```
